Adding large-step encoding for CHCs in CProver (no [required] changes to the solver)

yvizel opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Adding CHCs data structures and utilities for resolution-based large step encoding

Commit:
Adding CHCs data structures and utilities for resolution-based large step encoding

Commit:
Enabling large step encoding using a flag

Comparison operators for chcs

Clean up old code

Small fix

Extracting new chcs after transformation

Moving from std::set to std::unordered_set

Fixing multiple issues (compilation, formatting, etc.)

Fixing compilation errors

Amazon Linux 2023 compilation instructions

This updates the file COMPILING.md with instructions for compiling on Amazon
Linux 2023.

Add CI performance analysis script and workflow

Add scripts/ci_analysis.py for automated analysis of GitHub Actions CI
run durations. The script downloads logs via the gh CLI and produces
detailed breakdowns at job, step, suite, and individual test levels.

Supported log formats:
- ctest verbose output (CMake-based jobs)
- make parallel output with runner-ID prefixes (Linux, macOS)
- make serial output without prefixes (Windows VS 2022)
- validate-trace-xml-schema timestamp-based output

Features:
- Per-job wall-clock durations sorted by slowest first
- Per-step durations within each job
- Per-suite durations for both ctest and make test runners
- Individual test drill-down within slow suites (>60s)
- Cross-run aggregation with mean/min/max across N runs
- Markdown output (--markdown) for GitHub Actions job summaries
- Log caching to avoid re-downloading (--log-dir)
- Workflow identified by filename for resilience to renames

Usage:
  python3 scripts/ci_analysis.py                    # 5 recent runs
  python3 scripts/ci_analysis.py --runs 10 --top 30
  python3 scripts/ci_analysis.py --markdown         # for CI summaries
  python3 scripts/ci_analysis.py --log-dir /tmp/cache

A GitHub Actions workflow (.github/workflows/ci-analysis.yaml) runs the
script on every push to develop and publishes the results as a job
summary visible in the Actions tab.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

Fix Xen CI job by moving to GitHub mirror

git://xenbits.xen.org/xen.git appears to be working fine from my local
host, but maybe the (plaintext) git protocol is being blocked somewhere.
Use GitHub's mirror of the Xen project instead.

Enable parallel regression test execution for Windows VS 2022

Add a 'test-parallel-jobs' Makefile target to both regression/Makefile
and jbmc/regression/Makefile that uses make -j for parallelism instead
of GNU Parallel (which is unavailable on Windows runners).

The target creates a phony rule per test directory using the
$(addprefix parallel__,$(DIRS)) pattern, allowing make's built-in
job server to schedule directories in parallel.

The VS 2022 CI job is updated to use this target with -j$windows-vcpus,
replacing the sequential 'make test' invocation. This reduces test
wall-clock time from 40 minutes down to 15 minutes.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

Replace deprecated upload-release-asset with action-gh-release

This is the recommended replacement and is also what we are already
using successfully in, e.g., cbmc-viewer.

Fixes: #7280

Bump microsoft/setup-msbuild from 2 to 3

Bumps [microsoft/setup-msbuild](https://github.com/microsoft/setup-msbuild) from 2 to 3.
- [Release notes](https://github.com/microsoft/setup-msbuild/releases)
- [Commits](https://github.com/microsoft/setup-msbuild/compare/v2...v3)

---
updated-dependencies:
- dependency-name: microsoft/setup-msbuild
  dependency-version: '3'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Fixing formatting issues and compilation errors

Add output_filet utility class from hw-cbmc

Promote output_filet from hw-cbmc's ebmc to src/util for reuse across
CBMC. The class provides RAII management of output streams, mapping
"-" to stdout and opening files otherwise.

Use output_filet in show_vcc.cpp and show_program.cpp to replace
duplicated open-file-or-stdout boilerplate.

Add unit tests for file output, stdout mapping, and error handling.

Not yet refactored:
- solver_factory.cpp: open_outfile_and_check returns unique_ptr<ofstream>
  that gets moved into solver objects; converting would touch the solvert
  ownership model.
- cprover_parse_options.cpp: uses a different pattern (no "-" support,
  returns error code instead of throwing).

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

Fix out-of-bounds read in MC/DC coverage eval_expr

The eval_expr function in cover_instrument_mcdc.cpp dereferences the
result of std::map::find() without checking for end(). This happens
when values_of_atomic_exprs skips inconsistent expressions (where
signs.size() != 1), leaving them absent from the map. When eval_expr
later encounters such an expression, find() returns end() and the
dereference is undefined behavior (stack-buffer-overflow detected by
AddressSanitizer).

Fix by checking the iterator before dereferencing.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

Initialise all POD members of goto_symex_statet

The constructor did not take care of them and our unit tests exposed
that, at least within unit tests, we were accessing uninitialised
members.

Fix reference to temporary that has gone out of scope

Store language_mode by value instead of by reference in
goto_symex_statet. The copy is cheap (irep_idt is an interned string
index) and eliminates the dangling-reference problem at the root: the
class now owns its data, so callers need not worry about the lifetime
of the argument.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

Initialise message_handler in bv_refinement unit test

The message_handler member of bv_refinementt::infot was left
uninitialised, which is undefined behaviour when it is later
dereferenced. Set it to null_message_handler.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

Add CORE label to unit-xfail and java-unit-xfail tests

These tests were missing the CORE label, so they were excluded from
coverage collection (ctest -V -L CORE). This meant XFAIL test code
never executed during coverage runs, producing 0% coverage for any
changes to XFAIL tests.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

Amazon Linux 2023 compilation instructions

This updates the file COMPILING.md with instructions for compiling on Amazon
Linux 2023.

Add CI performance analysis script and workflow

Add scripts/ci_analysis.py for automated analysis of GitHub Actions CI
run durations. The script downloads logs via the gh CLI and produces
detailed breakdowns at job, step, suite, and individual test levels.

Supported log formats:
- ctest verbose output (CMake-based jobs)
- make parallel output with runner-ID prefixes (Linux, macOS)
- make serial output without prefixes (Windows VS 2022)
- validate-trace-xml-schema timestamp-based output

Features:
- Per-job wall-clock durations sorted by slowest first
- Per-step durations within each job
- Per-suite durations for both ctest and make test runners
- Individual test drill-down within slow suites (>60s)
- Cross-run aggregation with mean/min/max across N runs
- Markdown output (--markdown) for GitHub Actions job summaries
- Log caching to avoid re-downloading (--log-dir)
- Workflow identified by filename for resilience to renames

Usage:
  python3 scripts/ci_analysis.py                    # 5 recent runs
  python3 scripts/ci_analysis.py --runs 10 --top 30
  python3 scripts/ci_analysis.py --markdown         # for CI summaries
  python3 scripts/ci_analysis.py --log-dir /tmp/cache

A GitHub Actions workflow (.github/workflows/ci-analysis.yaml) runs the
script on every push to develop and publishes the results as a job
summary visible in the Actions tab.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

Fix Xen CI job by moving to GitHub mirror

git://xenbits.xen.org/xen.git appears to be working fine from my local
host, but maybe the (plaintext) git protocol is being blocked somewhere.
Use GitHub's mirror of the Xen project instead.

Enable parallel regression test execution for Windows VS 2022

Add a 'test-parallel-jobs' Makefile target to both regression/Makefile
and jbmc/regression/Makefile that uses make -j for parallelism instead
of GNU Parallel (which is unavailable on Windows runners).

The target creates a phony rule per test directory using the
$(addprefix parallel__,$(DIRS)) pattern, allowing make's built-in
job server to schedule directories in parallel.

The VS 2022 CI job is updated to use this target with -j$windows-vcpus,
replacing the sequential 'make test' invocation. This reduces test
wall-clock time from 40 minutes down to 15 minutes.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

Replace deprecated upload-release-asset with action-gh-release

This is the recommended replacement and is also what we are already
using successfully in, e.g., cbmc-viewer.

Fixes: #7280

Bump microsoft/setup-msbuild from 2 to 3

Bumps [microsoft/setup-msbuild](https://github.com/microsoft/setup-msbuild) from 2 to 3.
- [Release notes](https://github.com/microsoft/setup-msbuild/releases)
- [Commits](https://github.com/microsoft/setup-msbuild/compare/v2...v3)

---
updated-dependencies:
- dependency-name: microsoft/setup-msbuild
  dependency-version: '3'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

cbmc-incr-smt2: Add test suffix to avoid intermittent failures

With concurrently running tests we might have had z3 vs cvc5 test
execution scribble over each other, resulting in non-reproducible test
failures.

Commit:
Enabling large step encoding using a flag

Commit:
Comparison operators for chcs

Commit:
Clean up old code

Commit:
Small fix

Commit:
Extracting new chcs after transformation

tautschnig commented on src/cprover/cprover_parse_options.h:
```diff
@@ -25,8 +25,10 @@ Author: Daniel Kroening, dkr@amazon.com
+  "(large-step)"                                                               \
   "(verbose)"
 
+
```
Nit pick: no need for extra blank line.

tautschnig commented on src/cprover/chc_db.cpp:
```diff
@@ -0,0 +1,76 @@
+
+#include "chc_db.h"
+
+#include <iostream>
```
Is this a debugging left-over?

tautschnig commented on src/cprover/chc_db.cpp:
```diff
@@ -0,0 +1,76 @@
+
+  for (auto & sp : m_db.get_state_preds())
+  {
+    std::set<exprt> outgoing;
```
Only use `std::set` when ordering is important or needs to be stable, else use `std::unordered_set`.

tautschnig commented on src/cprover/chc_db.cpp:
```diff
@@ -0,0 +1,76 @@
+      continue;
+    }
+    exprt func = to_function_application_expr(*r.head()).function();
+    m_head_idx[func].insert(&r);
```
Please refrain from storing pointers into other data structures, this is prone to break (the code here really requires that `m_clauses` will never change once `build_indices` has been called. Either make copying cheap (like `irept` does), or else store the index into `m_clauses` (so that `m_clauses` can still be added to).

yvizel replied:
Basically you're not supposed to add clauses after building the indices. If you do, you need to rebuild.
Does it still require a change?

tautschnig replied:
I think we need to do better than "good intentions." We need some way to ensure this cannot go wrong.

tautschnig commented on src/cprover/chc_db.cpp:
```diff
@@ -0,0 +1,76 @@
+    m_head_idx[func].insert(&r);
+
+    std::vector<symbol_exprt> use;
+    r.used_relations(*this,std::back_inserter(use));
```
I'd suggest that `used_relations` just return a `std::vector<symbol_exprt>` (which is cheap with copy elision).

tautschnig commented on src/cprover/cutpoint_graph.h:
```diff
@@ -0,0 +1,169 @@
+
+class cutpoint
+{
+  friend class cutpoint_graph;
```
Why do we need `friend`here?

tautschnig commented on src/cprover/cutpoint_graph.h:
```diff
@@ -0,0 +1,169 @@
+  friend class cutpoint_graph;
+
+  const cutpoint_graph &m_graph;
+  unsigned m_id;
```
`unsigned` will hardly ever be the right type. Should it perhaps be `std::size_t`?

tautschnig commented on src/cprover/cutpoint_graph.h:
```diff
@@ -0,0 +1,169 @@
+  const goto_programt::instructiont & m_inst;
+
+  typedef std::vector<cutpoint_edge_ptr> EdgeVec;
+  EdgeVec m_succ;
```
Use snake_case rather than CamelCase.

tautschnig reviewed (commented):
A bunch of comments that each apply in several more places. More generally, however, it would be great to have comments. At bare minimum there should be high-level design comments in some place that explain how the various pieces fit together. Also, we need tests that demonstrate the new feature.

Commit:
Moving from std::set to std::unordered_set

Commit:
Naming conventions, unsigned -> size_t

Commit:
Renaming types and function names, and adjusments to changes in API

Commit:
Adjusting API changes

Commit:
Equivalent query clauses may represent different properties. Allow it in CHC db

Commit:
Adding regression tests for large step encoding of CHCs

Commit:
Some comments and minor modifications

Commit:
Removed whitespace

Commit:
Removed include

Commit:
Missing commit

Commit:
Removing redundant include, and fixing function arguments

tautschnig commented on regression/cprover/large_step/trace/trace_pointer3.desc:
```diff
@@ -0,0 +1,9 @@
+KNOWNBUG
```
Applies to this one as well as all other `KNOWNBUG` tests that you added: could you please add at the bottom:
```
--
Brief explanation as to what the bug is.
```
That is, add another `--` and then an explanatory text.

tautschnig commented on src/cprover/chc_db.cpp:
```diff
@@ -0,0 +1,114 @@
+chc_dbt::chc_sett chc_dbt::m_empty_set;
+std::unordered_set<exprt, irep_hash> chc_grapht::m_expr_empty_set;
+
+void horn_clauset::used_relations(chc_dbt &db, std::vector<symbol_exprt> & out) const
```
You'll want to run `clang-format` (or, rather, `git-clang-format`) to ensure all use of whitespace is in line with the repository rules.

tautschnig commented on src/cprover/chc_db.cpp:
```diff
@@ -0,0 +1,114 @@
+    exprt func = to_function_application_expr(*r.head()).function();
+    m_head_idx[func].insert(i);
+
+    std::vector<symbol_exprt> use;
+    r.used_relations(*this, use);
```
Please change the signature of `used_relations` such that this becomes `std::vector<symbol_exprt> use = r.used_relations(*this);`. That is, return by value.

tautschnig commented on src/cprover/chc_db.cpp:
```diff
@@ -0,0 +1,114 @@
+
+  std::unordered_set<function_application_exprt, irep_hash> funcs;
+  body->visit_pre([&funcs](const exprt &expr) {
+                    if (can_cast_expr<function_application_exprt>(expr))
+                    {
+                      const function_application_exprt & f = to_function_application_expr(expr);
+                      funcs.insert(f);
+                    }
```
```suggestion
                    if (auto f = expr_try_dynamic_cast<function_application_exprt>(expr))
                    {
                      funcs.insert(*f);
                    }
```

tautschnig commented on src/cprover/chc_db.h:
```diff
@@ -0,0 +1,233 @@
+  forall_exprt m_chc;
+
+public:
+  horn_clauset(forall_exprt f) : m_chc(f) {}
```
Qualify this with `explicit`

tautschnig commented on src/cprover/chc_db.h:
```diff
@@ -0,0 +1,233 @@
+    return m_chc;
+  }
+
+  const exprt* body() const {
+    if (can_cast_expr<implies_exprt>(m_chc.where()))
+    {
+      return &to_implies_expr(m_chc.where()).op0();
+    }
+    return &m_chc.where();
+  }
```
Why does this return a pointer and not just a reference?

tautschnig commented on src/cprover/chc_db.h:
```diff
@@ -0,0 +1,233 @@
+  bool is_fact() const {
+    auto b = body();
+    bool not_fact = false;
+    b->visit_pre(
+      [&not_fact](const exprt &expr)  {
+        if(can_cast_expr<function_application_exprt>(expr))
+        {
+          not_fact = true;
+        }
+      });
```
I think you'd be better off using `depth_iteratort` so that you can abort the traversal as soon as a you hit a `function_application_exprt`.

tautschnig commented on src/cprover/chc_db.h:
```diff
@@ -0,0 +1,233 @@
+    {
+      auto h = head();
+      bool res = true;
+      h->visit_pre(
+        [&res](const exprt &expr) {
+          if(can_cast_expr<function_application_exprt>(expr))
+            res = false;
+      });
```
As above: seems to be a case for `depth_iteratort`.

tautschnig commented on src/cprover/chc_db.h:
```diff
@@ -0,0 +1,233 @@
+    return !(*this==other);
+  }
+
+  bool operator<(const horn_clauset &other) const
+  {
+    return m_chc < other.m_chc;
+  }
```
When do you need this one? It's stable, but still rather arbitrary from a human's point of view.

tautschnig commented on src/cprover/chc_db.h:
```diff
@@ -0,0 +1,233 @@
+ */
+class chc_dbt
+{
+  friend class horn_clauset;
```
Where do you need the friendship?

tautschnig commented on src/cprover/chc_db.h:
```diff
@@ -0,0 +1,233 @@
+
+  void add_clause(const forall_exprt & f)
+  {
+    if (f.is_true())
+      return;
```
Can this ever be the case? `f` is a `forall_exprt` (and not a forall-expression-processed-by-the-simplifier).

tautschnig commented on src/cprover/chc_db.h:
```diff
@@ -0,0 +1,233 @@
+  static std::unordered_set<exprt, irep_hash> m_expr_empty_set;
+
+public:
+  chc_grapht(chc_dbt & db) : m_db(db), m_entry(nullptr) {}
```
Qualify with `explicit`.

tautschnig commented on src/cprover/cprover_parse_options.cpp:
```diff
@@ -277,6 +277,7 @@ int cprover_parse_optionst::main()
 
     solver_options.trace = cmdline.isset("trace");
     solver_options.verbose = cmdline.isset("verbose");
+    solver_options.large_step = cmdline.isset("large-step");
```
Please add a `help` entry (in `cprover_parse_optionst::help` at the bottom of this file).

tautschnig commented on src/cprover/cutpoint_graph.cpp:
```diff
@@ -0,0 +1,230 @@
+#include "cutpoint_graph.h"
+#include "language_util.h"
+
+#include <iostream>
+#include <fstream>
```
Are either of those really required?

tautschnig commented on src/cprover/cutpoint_graph.cpp:
```diff
@@ -0,0 +1,230 @@
+#include <iostream>
+#include <fstream>
+
+cutpoint_grapht::~cutpoint_grapht() {
+  m_edges.clear();
+  m_cps.clear();
+  m_insts.clear();
+}
```
Why is the explicit deconstruction required?

tautschnig commented on src/cprover/state_encoding.cpp:
```diff
@@ -1233,6 +1235,56 @@ void variable_encoding(
+
+  rv.populate_new_db();
+
+  container_encoding_targett container2;
+  std::vector<horn_clauset> all2;
```
What are those good for?

tautschnig commented on src/cprover/state_encoding.cpp:
```diff
@@ -1233,6 +1235,56 @@ void variable_encoding(
+{
+  chc_dbt db;
+  for (auto & clause : small_step_container.constraints) {
+    if (!can_cast_expr<forall_exprt>(clause))
+    {
+      throw incorrect_goto_program_exceptiont("Not forall");
+    }
+    const forall_exprt& forall = to_forall_expr(clause);
+    db.add_clause(forall);
```
I'd rewrite this using `expr_try_dynamic_cast` instead of `can_cast_expr`.

tautschnig reviewed (approved):
1. `clang-format` fixes are strictly required.
2. A bunch of suggestions below. Would be great to see them addressed, but we can do this in parallel with reviewing/working on the solver changes.

tautschnig commented:
Also, once done with the changes, please rebase/combine all commits into a single one. (Or perhaps one for the source changes and one for the tests.)

yvizel committed:
Changing user_relations function declaration

yvizel committed:
clang format, and minor fixes to is_query, is_fact

yvizel committed:
Small fix

yvizel committed:
Merge branch 'diffblue:develop' into develop

yvizel committed:
Merge branch 'diffblue:develop' into develop

yvizel committed:
Merge branch 'diffblue:develop' into develop

yvizel committed:
Merge branch 'diffblue:develop' into develop

yvizel committed:
feat(cprover): Add large-step encoding for CHCs

Introduce a resolution-based large-step encoding for Constrained Horn
Clauses (CHCs) in CProver. The encoding is opt-in via a new
`--large-step` command-line flag and does not require any changes to
the underlying solver.

New components added under src/cprover/:
- chc_db: data structures for storing and indexing CHCs
- chc_large_step: the large-step transformation built on top of chc_db
- chc_wto: weak topological ordering of CHC predicates
- cutpoint_graph: cutpoint graph used to drive the large-step encoding

Hooks the new option through cprover_parse_options and extends
state_encoding to emit the transformed clauses when the flag is set.

Adds regression tests under regression/cprover/large_step/ covering
arrays, pointers, structs, loops, function calls, and trace generation.

yvizel committed:
Merge branch 'diffblue:develop' into develop
